java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,555 @@
+    connect();
+  }
+
+  private void connect() throws IOException
+  {
+    if (socket != null)
+      return;
+
+    Socket newSocket = new Socket();
+    newSocket.connect(new InetSocketAddress(hostname, 7818));
+    newSocket.setSoTimeout(30000);
+    newSocket.setTcpNoDelay(true);
+    socket = newSocket;
+
+    outStream = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
+    inStream = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+
+    if (transcodeMode != null && transcodeMode.length() > 0)
+    {
+      outStream.write(("XCODE_SETUP " + transcodeMode + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      String response = Sage.readLineBytes(inStream);
+      if (!"OK".equals(response))
+        throw new java.io.IOException("Error with remote transcode setup for " + transcodeMode + " of: " + response);
+    }
+
+    if (readonly)
+      outStream.write("OPENW ".getBytes(Sage.BYTE_CHARSET));
+    else
+      outStream.write("WRITEOPENW ".getBytes(Sage.BYTE_CHARSET));
+
+    outStream.write(remoteFilename.getBytes(StandardCharsets.UTF_16BE));
+    outStream.write("\r\n".getBytes(Sage.BYTE_CHARSET));
+    outStream.flush();
+
+    String response = Sage.readLineBytes(inStream);
+    if (!"OK".equals(response))
+    {
+      // Don't leave the connection open. We can't read anything if we are seeing this error anyway.
+      disconnect();
+      throw new IOException("Error opening remote file of:" + response);
+    }
+  }
+
+  private void disconnect()
+  {
+    if (socket != null)
+    {
+      try
+      {
+        outStream.write("QUIT\r\n".getBytes(Sage.BYTE_CHARSET));
+        outStream.flush();
+      }
+      catch (IOException e) {}
+
+      try
+      {
+        socket.close();
+      }
+      catch (IOException e) {}
+      socket = null;
+    }
+    if (outStream != null)
+    {
+      try
+      {
+        outStream.close();
+      }
+      catch (IOException e) {}
+      outStream = null;
+    }
+    if (inStream != null)
+    {
+      try
+      {
+        inStream.close();
+      }
+      catch (IOException e) {}
+      inStream = null;
+    }
+  }
+
+  private synchronized void reconnect() throws IOException
+  {
+    disconnect();
+    connect();
+  }
+
+  private long getMaxRead(long position, long count) throws IOException
+  {
+    // We can't read more than what is in the file. The first part checks a cached variable, the
+    // second part checks the file directly. Unless the stream really lingers trying to get a
+    // massive read towards the end, this should be a very minimal load on the media server.
+    count = Math.min(
+        maxRemoteSize == 0 || (position + count >= maxRemoteSize && activeFile) ?
+            reallyGetLength() - position : maxRemoteSize - position,
+        count);
+
+    // If we are trying to read beyond the end of the file, don't read anything.
+    return (count == 0) ? 0 : count;
```

Comment:
That was changed a few times and the end result is like that because I didn't notice I had changed it into something obvious.